### PR TITLE
feat: Avoid Redundant SSA for Manifest Patching

### LINKED ISSUE #1456: Avoid ssa for Manifest CR when no update needed
### Description

At the moment, kyma controller will do ssa for all enabled modules, no matter if there is change or not, and in most cases, the related manifest CRs only need to be updated if there is update in the moduletemplate.

Solution:

Based on the research https://github.com/kyma-project/lifecycle-manager/issues/1481, we can use the cached manifest CR to avoid in cluster request to fetch the latest status

Kyma `status.modules[].version` tracking the current enabled manifest CR version, before doing ssa of manifest, we should compare it with the version of the current moduletemplate, only update manifest when the version get updated.

Include the check of comparing the `status.modules[].channel` as well.

https://github.com/kyma-project/lifecycle-manager/blob/bb2bf933d260730000b6e7f56b50fa0e5dfd0809/pkg/module/sync/runner.go#L121-L132

### Reasons

reduce requests to API server.

### Acceptance Criteria
- [x] Only do manifest CR update, by extending this [function](https://github.com/kyma-project/lifecycle-manager/blob/bb2bf933d260730000b6e7f56b50fa0e5dfd0809/pkg/module/sync/runner.go#L162), when: 
    - [x] There is new version for the moduletemplate 
    - [x] Or the channel has been switched 
    - [x] Or Manifest CR status.state has been changed compared to KymaCr.Status.Modules[x].state 
- [x] Cover scenario in tests.
- [x] Extend [this test](https://github.com/kyma-project/lifecycle-manager/blob/a0c49436f3d11d03c9a7556ec11c7c9f69d621d9/tests/e2e/module_upgrade_new_version_test.go#L12) to check Module Version  in `kyma.status` and ManifestCR gets updated in E2E

### Feature Testing

Unit Tests, check if possible for integration or E2E test

### Testing approach

_No response_

### Attachments

_No response_

### PR DESCRIPTION
**Pull request by LeelaChacha**
Closes #1456

### CODE REVIEW

**Comment by LeelaChacha** [comment]
NOTE: One integration test is failing as it is invalidated now. It is supposed to check that a manifest is reset if manually overwritten. However, now manifest patching only works when new template version, channel, or kyma module status is present.
How to proceed with the test? Should we reverse the check to show that manifest ssa is not?

**Review by c-pius — COMMENTED** [review]
Thanks for the re-work! Can confirm that the flickering is now gone 👍🏻 

> NOTE: One integration test is failing as it is invalidated now. It is supposed to check that a manifest is reset if manually overwritten. However, now manifest patching only works when new template version, channel, or kyma module status is present.
How to proceed with the test? Should we reverse the check to show that manifest ssa is not?

This is an important finding. We actually changed parts of our handling, hence the test fails. I also tested this manually. Before, when e.g. changing `Spec.Install.Source.Repo` of a ManifestCR, we reconciled it back to the original state. Now this is not happening anymore except we also change the version, or channel. I guess this would be acceptable as ManifestCRs reside in KCP and can only be changed by us or SRE. I still want to double check... @ruanxin is this something we explicitly accept?

**Comment by ruanxin** [comment]
> Thanks for the re-work! Can confirm that the flickering is now gone 👍🏻
> 
> > NOTE: One integration test is failing as it is invalidated now. It is supposed to check that a manifest is reset if manually overwritten. However, now manifest patching only works when new template version, channel, or kyma module status is present.
> > How to proceed with the test? Should we reverse the check to show that manifest ssa is not?
> 
> This is an important finding. We actually changed parts of our handling, hence the test fails. I also tested this manually. Before, when e.g. changing `Spec.Install.Source.Repo` of a ManifestCR, we reconciled it back to the original state. Now this is not happening anymore except we also change the version, or channel. I guess this would be acceptable as ManifestCRs reside in KCP and can only be changed by us or SRE. I still want to double check... @ruanxin is this something we explicitly accept?

Yes, normally modifying manifest CR spec is controlled by us, it's fine to skip this overwritten, however, it still has benefits that allow this overwritten.
one use case for troubleshooting:
1. add skip reconciliation label for one Kyma CR
2. modify related manifest CR spec for testing or troubleshooting
3. afterwards, remove skip reconciliation label
4. manifest CR get reset.

And it's possible to support this feature, just need to do a client diff check between the spec of manifest in cluster (`manifestInCluster`) and converted from moduletemplate (`manifestObj`), but we should monitor the kyma reconciliation time after release, to see how expensive this diff check.

**Comment by c-pius** [comment]
Thanks for chiming in! Just not sure if I understood you correctly since you explained both... Are you suggesting to accept it as proposed in this PR (don't overwrite manual changes), or are you suggesting to add the diff check as an additional evaluation in `NeedToUpdate` (and overwrite manual changes)?

**Comment by ruanxin** [comment]
> Thanks for chiming in! Just not sure if I understood you correctly since you explained both... Are you suggesting to accept it as proposed in this PR (don't overwrite manual changes), or are you suggesting to add the diff check as an additional evaluation in `NeedToUpdate` (and overwrite manual changes)?

I suggest adding this diff check in this PR, otherwise, we lose the support for that use case.

**Comment by c-pius** [comment]
> > Thanks for chiming in! Just not sure if I understood you correctly since you explained both... Are you suggesting to accept it as proposed in this PR (don't overwrite manual changes), or are you suggesting to add the diff check as an additional evaluation in `NeedToUpdate` (and overwrite manual changes)?
> 
> I suggest adding this diff check in this PR, otherwise, we lose the support for that use case.

Thanks for clarifying!

@LeelaChacha would you mind checking if this can be added to the `NeedToUpdate`?

**Comment by LeelaChacha** [comment]
@ruanxin @c-pius does a diff of the spec suffice?

**Comment by c-pius** [comment]
> @ruanxin @c-pius does a diff of the spec suffice?
> 
> Test is now green, btw

I guess checking the spec part should suffice. However, I am not sure if `reflect.DeepEqual` is the way to go. As mentioned [here](https://github.com/kubernetes-sigs/kubebuilder/issues/592), K8s may provide default values for unset fields. When doing so, it will be unequal all the time for certain manifests. Maybe this `equality.Semantic.DeepDerivative` is a good alternative. Would you mind checking with some debug comments whether it works?

And one other thing for @ruanxin. I vaguely remember that somebody mentioned that we transitioned to using Server Side Apply to transfer the compute load of comparing the resources from our controller to the API server (not sure if I remember this correctly). We are now re-adding this load on the controller. Are we sure we want to do that?

**Review by c-pius — APPROVED** [review]

File: internal/descriptor/provider/provider.go [hunk 1]
```diff
@@ -70,6 +71,20 @@ func (c *CachedDescriptorProvider) GetDescriptor(template *v1beta2.ModuleTemplat
...
+		return ErrTemplateNil
+	}
+	template.Spec.Descriptor.Object = descriptor
+	encoded, err := compdesc.Encode(descriptor.ComponentDescriptor, compdesc.DefaultJSONCodec)
+	template.Spec.Descriptor.Raw = encoded
+	if err != nil {
+		return errors.Join(ErrEncode, err)
+	}
+	return nil
+}
```
Comment (c-pius) [inline internal/descriptor/provider/provider.go:86]: This is production code that is only used for testing. Such things should be avoided in general (sometimes not possible though). Here is a proposal for how we can fix the integration test without adding new production code:

https://github.com/LeelaChacha/kyma-lifecycle-manager/pull/1

**Review by c-pius — CHANGES_REQUESTED** [review]

File: internal/descriptor/provider/provider.go (hunk 1 above)
Comment (LeelaChacha) [reply internal/descriptor/provider/provider.go]: that is a really clean solution! The only downside I see is the extra file, but objectively better than extra functions only for test.

**Review by c-pius — APPROVED** [review]

File: docs/technical-reference/api/manifest-cr.md [hunk 2]
```diff
@@ -2,6 +2,16 @@
...
 
+## Patching
+
+The [Runner](../../../pkg/module/sync/runner.go) is responsible for creating and updating manifests via Server Side Apply (SSA). An update is only performed when one of the following conditions is met:
+
+1. The version of the manifest object is different from the version in the Kyma module status.
+2. The channel of the manifest object is different from the channel in the Kyma module status.
+3. The state of the manifest object is different from the state in the Kyma module status.
+
+*NOTE:* The module status is not present in Kyma for mandatory modules, hence their manifest is updated via SSA in every reconcile loop.
```
Comment (mmitoraj) [inline docs/technical-reference/api/manifest-cr.md:13]:
```suggestion
>[!NOTE] 
>The module status is not present in Kyma for mandatory modules, hence their manifest is updated using SSA in every reconcile loop.
```
in Kyma? meaning Kyma CR?

File: docs/technical-reference/api/manifest-cr.md [hunk 3]
```diff
@@ -2,6 +2,16 @@
 
 The [Manifest custom resource (CR)](../../../api/v1beta2/manifest_types.go) is our internal representation of what results from the resolution of a ModuleTemplate CR in the context of a single cluster represented by a Kyma CR. Thus, a lot of configuration elements are similar or entirely equivalent to the data layer found in a ModuleTemplate CR.
 
+## Patching
+
+The [Runner](../../../pkg/module/sync/runner.go) is responsible for creating and updating manifests via Server Side Apply (SSA). An update is only performed when one of the following conditions is met:
+
+1. The version of the manifest object is different from the version in the Kyma module status.
```
Comment (mmitoraj) [inline docs/technical-reference/api/manifest-cr.md:9]:
```suggestion
1. The Manifest CR version differs from the Kyma CR's module status version.
```

File: docs/technical-reference/api/manifest-cr.md [hunk 4]
```diff
@@ -2,6 +2,16 @@
...
+2. The channel of the manifest object is different from the channel in the Kyma module status.
```
Comment (mmitoraj) [inline docs/technical-reference/api/manifest-cr.md:10]:
```suggestion
2. The Manifest CR channel differs from the Kyma CR's module status channel.
```

File: docs/technical-reference/api/manifest-cr.md [hunk 5]
```diff
@@ -2,6 +2,16 @@
...
+3. The state of the manifest object is different from the state in the Kyma module status.
```
Comment (mmitoraj) [inline docs/technical-reference/api/manifest-cr.md:11]:
```suggestion
3. The Manifest CR state differs from the Kyma CR's module status state.
```

**Review by c-pius — APPROVED** [review]

**Review by mmitoraj — APPROVED** [review]

**Pull request merged**